src/sentry/static/sentry/app/components/avatarCropper.tsx

```diff
@@ -0,0 +1,464 @@
+      pageY = mousePosition.pageY;
+    }
+
+    if (left < 0) {
+      left = 0;
+      pageX = mousePosition.pageX;
+    } else if (left + resizeDimensions.size > cropContainer.clientWidth) {
+      left = cropContainer.clientWidth - resizeDimensions.size;
+      pageX = mousePosition.pageX;
+    }
+
+    this.setState(state => ({
+      resizeDimensions: {...state.resizeDimensions, top, left},
+      mousePosition: {pageX, pageY},
+    }));
+  };
+
+  onMouseDown = (ev: React.MouseEvent<HTMLDivElement>) => {
+    ev.preventDefault();
+    this.setState({mousePosition: {pageY: ev.pageY, pageX: ev.pageX}});
+
+    document.addEventListener('mousemove', this.updateDimensions);
+    document.addEventListener('mouseup', this.onMouseUp);
+  };
+
+  onMouseUp = (ev: MouseEvent) => {
+    ev.preventDefault();
+    document.removeEventListener('mousemove', this.updateDimensions);
+    document.removeEventListener('mouseup', this.onMouseUp);
+    this.drawToCanvas();
+  };
+
+  startResize = (direction: Position, ev: React.MouseEvent<HTMLDivElement>) => {
+    ev.stopPropagation();
+    ev.preventDefault();
+    document.addEventListener('mousemove', this.updateSize);
+    document.addEventListener('mouseup', this.stopResize);
+
+    this.setState({
+      resizeDirection: direction,
+      mousePosition: {pageY: ev.pageY, pageX: ev.pageX},
+    });
+  };
+
+  stopResize = (ev: MouseEvent) => {
+    ev.stopPropagation();
+    ev.preventDefault();
+    document.removeEventListener('mousemove', this.updateSize);
+    document.removeEventListener('mouseup', this.stopResize);
+
+    this.setState({resizeDirection: null});
+    this.drawToCanvas();
+  };
+
+  updateSize = (ev: MouseEvent) => {
+    const cropContainer = this.cropContainer.current;
+    if (!cropContainer) {
+      return;
+    }
+
+    const {mousePosition} = this.state;
+
+    const yDiff = ev.pageY - mousePosition.pageY;
+    const xDiff = ev.pageX - mousePosition.pageX;
+
+    this.setState({
+      resizeDimensions: this.getNewDimensions(cropContainer, yDiff, xDiff),
+      mousePosition: {pageX: ev.pageX, pageY: ev.pageY},
+    });
+  };
+
+  // Normalize diff across dimensions so that negative diffs are always making
+  // the cropper smaller and positive ones are making the cropper larger
+  getDiffNW = (yDiff: number, xDiff: number) => {
+    return (yDiff - yDiff * 2 + (xDiff - xDiff * 2)) / 2;
+  };
+
+  getDiffNE = (yDiff: number, xDiff: number) => {
+    return (yDiff - yDiff * 2 + xDiff) / 2;
+  };
+
+  getDiffSW = (yDiff: number, xDiff: number) => {
+    return (yDiff + (xDiff - xDiff * 2)) / 2;
+  };
+
+  getDiffSE = (yDiff: number, xDiff: number) => {
+    return (yDiff + xDiff) / 2;
+  };
+
+  getNewDimensions = (container: HTMLDivElement, yDiff: number, xDiff: number) => {
+    const {resizeDimensions: oldDimensions, resizeDirection} = this.state;
+
+    const diff = this['getDiff' + resizeDirection!.toUpperCase()](yDiff, xDiff);
```

Comment:
@EvanPurkhiser Is the `!` a mistake? I don't know what it means in this context.

Reply:
It's a non-null assertion operator. Basically it tells typescript resizeDirection is NOT null. In thise case we know it's not null because the only context where `getNewDimensions` is called is while we're resizing the cropper, in which case direction will be known.

I could do a check above, but this function needs a consistent return.

Reply:
@EvanPurkhiser Got it, makes sense. Though if `resizeDirection` is `null`, couldn't you just return the original dimensions? Granted, based off the original code this should never happen.

Reply:
Yeah true I could do that. I'm fine with not though. If you feel strongly I can change it.

Reply:
@EvanPurkhiser I'm fine with it as is, just pointing it out :)